### PR TITLE
Fix preloader softlocks caused by bar-scaling animation bugs

### LINKED ISSUE #1214: "Cirno's Math School" Flash movie by IOSYS crashes instantly
A copy of the swf can be gotten here:
http://b2.aspbooru.de/Post/69/file

In Adobe Flash it works fine, but using ruffle with the current (as of 21 september) nightly it panics.

```
# Error Info
Error name: Error
Error message: Unknown panic
Error stack:
Y@http://b2.aspbooru.de/js/core.ruffle.d1fd7af53eba2318a50c.js:1:31241
__wbg_new_df6e6ab7a65c4c4d@http://b2.aspbooru.de/js/ruffle.js:1:808
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[1402]:0x22dde6
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[1304]:0x21ef9d
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[3466]:0x2c1691
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[2872]:0x2ae15c
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[3638]:0x2c46b2
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[3765]:0x2c620d
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[3241]:0x2bc356
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[406]:0x107aa2
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[2339]:0x28dc3e
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[932]:0x1d6e20
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[274]:0x553f4
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[1231]:0x213028
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[724]:0x19be93
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[292]:0x8b4c0
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[574]:0x163bb2
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[1409]:0x22eed0
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[2593]:0x29e945
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[2340]:0x28dd2e
@http://b2.aspbooru.de/js/1f537abbd497c48eaa79.module.wasm:wasm-function[3901]:0x2c774e
F@http://b2.aspbooru.de/js/core.ruffle.d1fd7af53eba2318a50c.js:1:30095
f@http://b2.aspbooru.de/js/core.ruffle.d1fd7af53eba2318a50c.js:1:29775

# Player Info
Allows script access: true
Player type: Object
SWF URL: /img/images/3b/dc/3bdc7f8bbfb648e825f79dbe6095288871b1780b1d37f8cd2ad4c50cae40b5ca.swf
Param movie: /img/images/3b/dc/3bdc7f8bbfb648e825f79dbe6095288871b1780b1d37f8cd2ad4c50cae40b5ca.swf
Param allowFullScreen: true
Attribute 0: [object Attr]
Attribute 1: [object Attr]
Attribute 2: [object Attr]

# Page Info
Page URL: http://b2.aspbooru.de/Post/69

# Browser Info
Useragent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:80.0) Gecko/20100101 Firefox/80.0
OS: Win32

# Ruffle Info
Ruffle version: 0.1.0
Ruffle source: undefined
```

### PR DESCRIPTION
**Pull request by kmeisthax**
Fixes #1214, #367 and maybe more issues involving preloader softlocks but I haven't checked the entire issue queue for it.

The underlying problem as discovered by @adrian17 is that `xscale` and `yscale` are stored as percentages in Flash Player, but unit proportions (e.g. 0 to 1) in Ruffle. This means accumulating a total in them (e.g. for animations) causes rounding errors. I changed `DisplayObjectBase` to always store scales as `Percent` (for unit bookkeeping) and then convert to unit proportions for all the matrix math display objects have to do. This doesn't appear to have any immediately obvious performance issues, and it fixes the bug.

### CODE REVIEW

File: core/src/lib.rs [hunk 1]
```diff
@@ -28,6 +28,7 @@ mod font;
 mod html;
 mod library;
 pub mod loader;
+mod percentage;
```
Comment (Herschel) [inline core/src/lib.rs:31]: Nit: let's name the module `percent` too

**Review by Herschel — CHANGES_REQUESTED** [review]
Thank you! Could you add a test for this too? Something as simple as:
```
_xscale = 12;
_xscale += 2;
trace(_xscale);
```

`_rotation` probably has similar issues. `_rotation` is in degrees but gets transformed to radians ASAP in the setter and then cached, but instead should be cached as degrees.  (If you do add a Degrees type in this PR, maybe we should put them both in a `types` module instead of having separate modules)

`_alpha` is also a percentage (although it isn't cached, so this specific issue doesn't happen there, and in fact, in some future PR, all of ColorTransform needs to eventually turn into fixed88, #193).

FYI for your own sanity, these are different in AS3 land; `scaleX`, `scaleY`, and `alpha` are all normalized 0-1 and are cached as such; scaling by 100 anywhere will result in the same kind of precision issues. So there will eventually have to be a conditional in here based on the SWF AS version. (`rotation` is still in degrees.)

File: core/src/lib.rs (hunk 1 above)
Comment (kmeisthax) [reply core/src/lib.rs:31]: Mooted: I'm naming it `types` instead and adding a `Degrees` type.

**Comment by kmeisthax** [comment]
I've added a test that checks `_xscale` and `_yscale` both - it turns out you need about 30% or so of range in order to actually trip this check, so I checked about 200 steps of 2. I also added `_rotation` and a `Degrees` type to cache these properties in their correct unit.

Thanks for the tip about AS3 - I'll make sure to write an AS3 version of this test when the time comes so that it can have it's angles and scales stored correctly, too.

**Comment by Herschel** [comment]
Thx!

**Pull request merged**